Makefile

```diff
@@ -507,24 +502,21 @@ KBUILD_CFLAGS   := -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs \
 		   -Werror=return-type -Wno-format-security \
 		   -std=gnu89
 KBUILD_CPPFLAGS := -D__KERNEL__
-KBUILD_RUSTCFLAGS :=
-# TODO: a simple way to update `Cargo.lock` when we add a new driver
-KBUILD_CARGOFLAGS := $(CARGO_VERBOSE) --locked \
-		    -Z build-std=core,alloc -Z unstable-options \
-		    --target=$(srctree)/arch/$(SRCARCH)/rust/target.json
+KBUILD_RUSTCFLAGS := --emit=dep-info,obj,link -Z binary_dep_depinfo=y \
```

Comment:
We should provide a way to add optional RUSTCFLAGS. IIRC this would override any we set from outside?

rust flags would be needed to set a custom rust sysroot when the normal sysroot in ro

Reply:
I don't recall having tested it recently, but the user should be able to pass `KRUSTCFLAGS`.

Reply:
Tried this, but it failed:

~~~sh
[nix-shell:/persist/data/kloenk/proj/linux]$ make O=build BINDGEN=$(readlink -f ./result/bin/bindgen) CC=clang KBUILD_RUSTCFLAGS="--sysroot=$(readlink -f ../sysroot) --emit=dep-info,obj,link -Z binary_dep_depinfo=y -C panic=abort -C embed-bitcode=n -C lto=n -C rpath=n -C force-unwind-tables=n -C codegen-units=1 -Z symbol-mangling-version=v0"
make[1]: Entering directory '/persist/data/kloenk/proj/linux/build'
  GEN     Makefile
  CALL    ../scripts/checksyscalls.sh
  CALL    ../scripts/atomic/check-atomics.sh
  RUSTC P rust/libmodule.so
make[2]: *** No rule to make target '/home/kloenk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/lib.rs', needed by 'rust/core.o'.  Stop.
make[1]: *** [/persist/data/kloenk/proj/linux/Makefile:1265: prepare0] Error 2
make[1]: Leaving directory '/persist/data/kloenk/proj/linux/build'
make: *** [Makefile:185: __sub-make] Error 2
~~~
the absolute path of `../sysroot` is `/persist/data/kloenk/proj/sysroot`. so It should never have picked up anything under `/home`

Reply:
same error with `KRUSTCFLAGS`. not sure why rustc does not use the sysroot

Reply:
> I don't recall having tested it recently, but the user should be able to pass `KRUSTCFLAGS` (like `KCFLAGS` for C etc.).

Although `rustc` is picky about passing a flag several times to override it, so it should work for new flags, but probably not for overriding the sysroot...

Reply:
> so It should never have picked up anything under `/home`

Yeah, the sysroot is fixed because I just call `rustc --print sysroot` without passing the flags. From a quick try that output changes by passing `--sysroot`, so I can just pass the flags. Let me quickly change it and you may try again.

Reply:
Give it a try now if you want.

Reply:
sadly that did not help

Reply:
Can you take a look at the `V=1` output to see what breaks?

Reply:
cmd: `make O=build BINDGEN=$(readlink -f ./result/bin/bindgen) CC=clang KRUSTCFLAGS=--sysroot=$(readlink -f ../sysroot) V=1`
stdout: https://termbin.com/3amr0
stderr:
~~~
make[2]: *** No rule to make target '/home/kloenk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/lib.rs', needed by 'rust/core.o'.  Stop.
make[1]: *** [/persist/data/kloenk/proj/linux/Makefile:1265: prepare0] Error 2
make: *** [Makefile:185: __sub-make] Error 2
~~~

Reply:
It does not make sense. `--sysroot` is there in the flags when you build another file:

      rustc ... --sysroot=/persist/data/kloenk/proj/sysroot ...

which means `rustc_sysroot` should be that one. I have also tried using sysroot myself:

    make LLVM=1 KRUSTCFLAGS='--sysroot /my/sysroot'

which works fine. Can you please double check that you had checked out the latest commit?

Also, just to be sure we are not going crazy, what is the output of

    rustc --sysroot TEST --print sysroot

? (it should be `TEST`)

Reply:
my checkout is `9ea67cb79645ec1cf41a5e96498c0e0b68b63bf6`. I only use CC=clang and no lld.ld as I don't have that installed at the moment. the rust --sysroot ... test prints test (so it succeeds). My rustc version is `rustc 1.51.0-nightly (6184f2395 2021-01-09)`.
I also did rm -r build to make sure there is nothing left of an older version.

Reply:
OK, fixed -- can you please give it a quick try to confirm?

Reply:
on it, already found this problem: https://github.com/Rust-for-Linux/linux/pull/52/files#r559441439

Reply:
What was the fix?

Reply:
Getting the sysroot was working in the body of the rules (which is why `--sysroot` was there in the command line in your output), but not on the target prerrequisites because of the order Kbuild includes the involved Makefiles. So I have now make second-expanding it, which means `rustc_flags` is now there when executing the command. I have yet to write a CI test for this.

All your tests for NixOS plus the `O=` ones etc. have been very helpful to make this work as smooth as possible for future users, so thanks a lot!

Reply:
It build and booted with `CONFIG_RUST_EXAMPLE=y`. Not tested anything else, but It should work I guess.

Reply:
And `CONFIG_RUST_EXAMPLE=m` also works.
Great job, Thanks

Reply:
Thanks to you! :-)

Reply:
> I have yet to write a CI test for this

CI test added in v7.